### DISCUSSION TITLE
Use counter with rotary encoder?

### DISCUSSION THREAD

**Discussion by Bubbleman75** [arendst/Tasmota#16002] [Q&A] [answered]
Hello

I want to use a ESP8266 (Wemos D1 Mini) and a rotary encoder (with some mechanics) to measure the water level in my cistern. The value is to be transmitted via MQTT.
When I configure GPIOs to RotaryA/RotaryB, the resolution (Dimmer: 0-100, 100 steps) is to low. I need about 3000 steps (0-300 cm)
Can I use the counter function in conjunction with an rotary encoder? The counter would have to count up AND down depending on the direction of rotation. And I need to be able to set the counter, from time to time, to a certain value, to calibrate the measurment.
Anyone have any advice on how I can best make this work?

Thanks a lot
Thomas

**Comment by sfromis** [comment, marked as answer]
`SetOption43` allows you to config rotary steps, up to 255. Not what you ask for. It is designed for a "volume control" (typically for a lamp) with a moderate range.

If you instead do a `SetOption43 1`, restricting the rotary range to 0/1, position 1 always means increment, and position 0 always means decrement. This can work with simple rules where you can add/subtract a counter value without the "volume control" logic.

**Comment by Bubbleman75** [comment]
Hello sfromis
I tried SetOption43 1. Results are 0 or 100 !?
```
10:17:50.608 MQT: Gartensensor2/stat/RESULT = {"POWER":"OFF","Dimmer":0}
10:17:53.486 MQT: Gartensensor2/stat/RESULT = {"POWER":"OFF","Dimmer":100}
```
But that should be OK.

But I need some debounce. When I turn the knob a little faster, thera are incorrect values
MQTT Explorer:
![grafik](https://user-images.githubusercontent.com/5702279/180599006-7162dc8a-910d-4014-80dd-271e04a46116.png)
![grafik](https://user-images.githubusercontent.com/5702279/180599017-b374c3de-f8ae-476e-8a69-7180607b5e94.png)

Is it possible to debounce a rotary encoder?

Thomas

**Reply by sfromis** [reply]
What I suggested was to use rules and position to get the rotary signals, **not** using it as light control and dimmer. This is enabled using `SetOption98` and rules like: 
`on Rotary1#Pos1==1 do var1 Increment endon on Rotary1#Pos1==0 do var1 Decrement endon`

The rotary driver has a build time adjustable default of 
`#define ROTARY_DEBOUNCE      10                // Debounce time in milliseconds
`

**Comment by BastiJoe** [comment]
Maybe we can kill two birds with one stone here:
#17673

**Comment by brettlg** [comment]
Hi All,
I must be missing something obvious. I can get rotary encoder working with dimmer but I cannot for the life of me get it to dim to 0 (off) by turning the knob. I want to use the push button on my encoder for another purpose so wanting to dim to off with the rotary encoder. Minimum I can get is 1. Also, just a little thing but the change to dim level is not showing in the GUI without refresh, it does however show in console. Ideally I would like dimmer range to be 0 or 10-100% as the LEDs don't perform well under 10%. My rules are as follows...

Rule2 on Rotary2#Pos1==1 do dimmer + endon on Rotary2#Pos1==0 do dimmer - endon

Any help much appreciated
Thanks

**Reply by sfromis** [reply]
Do not cross-post your question in unrelated threads.

**Reply by brettlg** [reply]
My Apologizes
